Add documentation for Kubernetes health checks (#60201)
gravitational/teleport: docs/pages/enroll-resources/kubernetes-access/health-checks.mdx

---
title: Teleport Kubernetes Health Checks
sidebar_label: Health Checks
description: How to configure Teleport Kubernetes health checks and view health.
tags:
- conceptual
- zero-trust
- infrastructure-identity
---

> This documentation provides an overview of health checks to Kubernetes clusters with Teleport. Teleport Kubernetes Services periodically check the connectivity and permissions of enrolled Kubernetes clusters, and is available in Teleport version `18.3.0` and later.

Reviewer: ```suggestion
This documentation provides an overview of health checks to Kubernetes clusters with Teleport. Teleport Kubernetes Service instances periodically check the connectivity and permissions of enrolled Kubernetes clusters, and is available in Teleport version `18.3.0` and later.
```

Reviewer: probably should just `sed 's#Kubernetes Services#Kubernetes Service instances#g`

<Admonition type="warning">

`unknown` Kubernetes clusters may be unhealthy.

`unknown` Kubernetes clusters are not checked for health due to:
- Running a pre-`18.3.0` version of Teleport Kubernetes Service
- Explicitly configuring `health_check_config` labels to exclude a Kubernetes cluster
</Admonition>

## Viewing health

Kubernetes cluster health is viewed through the Teleport web UI, desktop Connect UI, `tctl` CLI tool, or Prometheus metrics.

**Teleport Web & Connect UI**

Click the refresh icon to get the latest health of Kubernetes clusters.

![Kubernetes health warning in the UI](../../../img/resource-health-check/kubernetes-health-warning.png)

**Teleport `tctl` CLI**

Run `tctl get kube_server/<your-kube-server-name>` for an overview of Kubernetes cluster health for a specific Kubernetes service.
```yaml
kind: kube_server
...
status:
  target_health:
    address: 192.168.106.2:58458
    message: 1 health check passed
    protocol: http
    status: healthy
    transition_reason: threshold_reached
    transition_timestamp: "2025-10-13T19:26:58.842855Z"
version: v3
```

Teleport Prometheus metrics

> 

Reviewer: We should note that metrics are disabled by default.
I would suggest including either a link or partial.

This partial explains pretty well:
`(!docs/pages/includes/diagnostics/diag-addr-prereqs-tabs.mdx!)`

(In fact we should probably add that to the reference page for metrics, because I had to dig into our "zero trust access" self-hosted metrics instructions to find it.)

**`tctl` health check commands**

Read the default health check config with `tctl get`:
```bash
$ tctl get health_check_config/default_kube
```

Create a new health check config with `tctl create`:
```bash
$ tctl create health_check_config.yaml
```

Update an existing config interactively with `tctl edit`:
```bash
$ tctl edit health_check_config/default_kube
```

Delete a health check config with `tctl rm`:
```bash
$ tctl rm health_check_config/example
```

Teleport Kubernetes Services are notified of changes to `health_check_config`, and reevaluate whether a Kubernetes cluster participates in health checks, applying any changes.

## FAQ

### How do I see unhealthy Kubernetes clusters?

The Teleport web and Connect UI highlight unhealthy Kubernetes clusters.

Clicking a highlighted Kubernetes cluster shows details of an unhealthy Kubernetes cluster.

![Kubernetes health warning in the UI](../../../img/resource-health-check/kubernetes-health-warning.png)

> It may take approximately `5m` for a health change to be reported.

Author: Good point, added the TODO.

Click the circular arrow refresh icon to get the latest health status. 

The Teleport `tctl` CLI tool also searches and displays unhealthy Kubernetes clusters.
```code
tctl kube ls --query 'health.status == "unhealthy"' 
```

### What guidance is there for troubleshooting unhealthy Kubernetes clusters?

See the [Kubernetes Service troubleshooting guide](./troubleshooting.mdx) for specific errors returned by health checks.

### How do I disable Kubernetes health checks?

> Add a filtering label to exclude Kubernetes clusters with the CLI command `tctl edit health_check_config/default_kube`. If you have multiple `health_check_config`, such as `health_check_config/example-a`, `health_check_config/example-b`, each config would be adjusted.

Reviewer: might want to drop a TODO comment if you're planning to add the `disabled/enabled` field we talked about as recommended way to disable a config

---
title: Teleport Kubernetes Health Checks
sidebar_label: Health Checks
description: How to configure Teleport Kubernetes health checks and view health.
tags:
- conceptual
- zero-trust
- infrastructure-identity
---

This documentation provides an overview of health checks to Kubernetes clusters with Teleport. Teleport Kubernetes Service instances periodically check the connectivity and permissions of enrolled Kubernetes clusters, and is available in Teleport version (= kubernetes.health_check_min_version =) and later.

> ## Why check the health of a Kubernetes cluster?

Reviewer 2: I think we can incorporate this section with the intro section so the user can get past the high-level information and to the substantial details more quickly.

- **Observability**: Discover network and permission issues before users do. Unhealthy Kubernetes clusters are visible in a Teleport UI, command line tool, or Prometheus metrics.
- **High Availability**: Automatically route and distribute connections to healthy Kubernetes clusters in a high-availability configuration.

> ## What's checked?

Reviewer 2: ```suggestion
## How Kubernetes health checks work
```
I think we can be a bit more descriptive about the content of this H2.

Reviewer 2: On a related note, we might want to combine "What's checked" and "Health states" into an H2-level section about how health checks work in order to group topics into a clearer hierarchy.

> Kubernetes permissions and a Kubernetes health endpoint are checked to determine whether a Kubernetes cluster is both up and usable with Teleport.

Reviewer 2: Since this is a conceptual and architectural guide, I would replace passive-voice constructions like "are checked" with active ones, e.g., "The Teleport Kubernetes Service checks...". This way, it's clear what Teleport components are responsible for what.

- `teleport_resources_health_status_healthy{type="kubernetes"}` is the total number of _healthy_ Kubernetes clusters
- `teleport_resources_health_status_unhealthy{type="kubernetes"}` is the total number of _unhealthy_ Kubernetes clusters
- `teleport_resources_health_status_unknown{type="kubernetes"}` is the total number of Kubernetes clusters in an _unknown_ state

A [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) expression may be used to determine the total number of Kubernetes clusters.
```promql
teleport_resources_health_status_healthy{type="kubernetes"} +
teleport_resources_health_status_unhealthy{type="kubernetes"} +
teleport_resources_health_status_unknown{type="kubernetes"}
```

A PromQL expression may be used to detect the presence of unhealthy Kubernetes clusters.
```promql
teleport_resources_health_status_unhealthy{type="kubernetes"} > 0
```

<Admonition type="note">
Prometheus metrics don't distinguish which cluster is unhealthy.

> Use a Teleport UI, or `tctl kube ls --query 'health.status == "unhealthy"'` to determine which cluster is unhealthy.

Reviewer 2: Does a Teleport UI refer to Teleport Connect and the Web UI? I would name these UIs explicitly.

I would also use a code fence with the `code` label and `$` prompt in front of the command to render our custom code component:

````suggestion
To determine which cluster is unhealthy, use the Teleport Web UI, Teleport Connect, or the following `tctl` command:

```code
$ tctl kube ls --query 'health.status == "unhealthy"'
```
````

  match:
    kubernetes_labels:
    - name: '*'
      values:
      - '*'
version: v1
```

A different `default` `health_check_config` also exists, and focuses on matching databases for health checks.

Multiple different `health_check_config` resources may be created for different groups of Kubernetes clusters. When multiple `health_check_config` match the same Kubernetes cluster, configs are sorted in ascending order by name, and only the first config applies (e.g., the name "00-my-config" has greater precedence than "10-my-config").

**`tctl` health check commands**

Read the default health check config with `tctl get`:
```bash
$ tctl get health_check_config/default_kube
```

Create a new health check config with `tctl create`:
```bash
$ tctl create health_check_config.yaml
```

Update an existing config interactively with `tctl edit`:
```bash
$ tctl edit health_check_config/default_kube
```

Delete a health check config with `tctl rm`:
```bash
$ tctl rm health_check_config/example
```

Teleport Kubernetes Service instances are notified of changes to `health_check_config`, and reevaluate whether a Kubernetes cluster participates in health checks, applying any changes.

## FAQ

> ### How do I see unhealthy Kubernetes clusters?

Author: Yes, it's moved into "Viewing health".

The Teleport web and Connect UI highlight unhealthy Kubernetes clusters.

Clicking a highlighted Kubernetes cluster shows details of an unhealthy Kubernetes cluster.

![Kubernetes health warning in the UI](../../../img/resource-health-check/kubernetes-health-warning.png)

It may take approximately `5m` for a health change to be reported.
{/* TODO(gavin): delete this line when we fix health status reporting delay */}

Click the circular arrow refresh icon to get the latest health status. 

The Teleport `tctl` CLI tool also searches and displays unhealthy Kubernetes clusters.
```code
tctl kube ls --query 'health.status == "unhealthy"' 
```

### What guidance is there for troubleshooting unhealthy Kubernetes clusters?

See the [Kubernetes Service troubleshooting guide](./troubleshooting.mdx) for specific errors returned by health checks.

> ### How do I disable Kubernetes health checks?

Reviewer 2: Should this content be in the "Configuring health checks" H2?

---
title: Teleport Kubernetes Health Checks
sidebar_label: Health Checks
description: How to configure Teleport Kubernetes health checks and view health.
tags:
- conceptual
- zero-trust
- infrastructure-identity
---

> This documentation provides an overview of health checks to Kubernetes clusters with Teleport. Teleport Kubernetes Services periodically check the connectivity and permissions of enrolled Kubernetes clusters, and is available in Teleport version `18.4` and later.

Reviewer 3: ```suggestion
This documentation provides an overview of health checks to Kubernetes clusters with Teleport. Teleport Kubernetes Services periodically check the connectivity and permissions of enrolled Kubernetes clusters, and is available in Teleport version `18.3.0` and later.
```

A [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) expression may be used to determine the total number of Kubernetes clusters.
```promql
teleport_resources_health_status_healthy{type="kubernetes"} +
teleport_resources_health_status_unhealthy{type="kubernetes"} +
teleport_resources_health_status_unknown{type="kubernetes"}
```

A PromQL expression may be used to detect the presence of unhealthy Kubernetes clusters.
```promql
teleport_resources_health_status_unhealthy{type="kubernetes"} > 0
```

<Admonition type="note">
Notice that Prometheus metrics don't distinguish which cluster is unhealthy.

Use a Teleport UI, or `tctl kube ls --query 'health.status == "unhealthy"'` to determine which cluster is unhealthy.
</Admonition>

The Teleport `tctl top` command displays Kubernetes health check metrics.

Steps to view health check metrics with `tctl`:
- Run `tctl top`
- Navigate to the "Raw Metrics" tab by pressing the right arrow `→` key several times
- Enter search mode by pressing the forward slash `/` key
- Type or paste `teleport_resources_health_status`

> ![Health metrics with the tctl top command](../../../img/resource-health-check/kubernetes-health-metrics-tctl.png)

Reviewer 3: I think we should omit the reference to `tctl top` here and focus on the UI and prometheus metrics.

  # Kubernetes clusters are matched by label selectors and at least one label selector
  # must be set.
  # If multiple `health_check_config` resources match the same Kubernetes cluster,
  # the matching health check configs are sorted by name and only the first
  # config applies.
  match:
    # kubernetes_labels matches Kubernetes cluster labels. An empty value is ignored.
    # If kubernetes_labels_expression is also set, then the match result is the logical
    # AND of both.
    kubernetes_labels:
      - name: env
        values:
          - dev
          - staging
    # kubernetes_labels_expression is a label predicate expression to match Kubernetes clusters.
    # An empty value is ignored.
    # If kubernetes_labels is also set, then the match result is the logical AND of both.
    kubernetes_labels_expression: 'labels["owner"] == "platform-team"'
```

The default `health_check_config` enables all Kubernetes clusters to participate in health checks from version `18.4` onward.
```yaml
kind: health_check_config
metadata:
  description: Enables all health checks by default
  labels:
    teleport.internal/resource-type: preset
  name: default

>   namespace: default

Reviewer 3: ```suggestion
  name: default_kube
```

```bash
$ tctl edit health_check_config/default
```

Delete a health check config with `tctl rm`:
```bash
$ tctl rm health_check_config/example
```

Teleport Kubernetes Services are notified of changes to `health_check_config`, and reevaluate whether a Kubernetes cluster participates in health checks, applying any changes.

## FAQ

### How do I see unhealthy Kubernetes clusters?

The Teleport web and Connect UI highlight unhealthy Kubernetes clusters.

Clicking a highlighted Kubernetes cluster shows details of an unhealthy Kubernetes cluster.

![Kubernetes health warning in the UI](../../../img/resource-health-check/kubernetes-health-warning.png)

It may take approximately `5m` for a health change to be reported.

Click the circular arrow refresh icon to get the latest health status. 

The Teleport `tctl` CLI tool also searches and displays unhealthy Kubernetes clusters.
```code
tctl kube ls --query 'health.status == "unhealthy"' 
```

### What guidance is there for troubleshooting unhealthy Kubernetes clusters?

See the [Kubernetes Service troubleshooting guide](./troubleshooting.mdx) for specific errors returned by health checks.

### How do I disable Kubernetes health checks?

> Add a filtering label to exclude Kubernetes clusters with the CLI command `tctl edit health_check_config/default`. If you have multiple `health_check_config`, such as `health_check_config/example-a`, `health_check_config/example-b`, each config would be adjusted.

Reviewer 3: ```suggestion
Add a filtering label to exclude Kubernetes clusters with the CLI command `tctl edit health_check_config/default_kube`. If you have multiple `health_check_config`, such as `health_check_config/example-a`, `health_check_config/example-b`, each config would be adjusted.
```

Adding any label to `health_check_config` tells Teleport to filter out Kubernetes clusters which don't match the label. 

For example, adding `healthcheck` and `enabled` to `kubernetes_labels` in the following example would exclude Kubernetes clusters without the label.
```yaml
kind: health_check_config
metadata:
  description: Enables all health checks by default
  labels:
    teleport.internal/resource-type: preset
  name: default

>   namespace: default

Reviewer 3: ```suggestion
  name: default_kube
```
